src/ess/bifrost/single_crystal/q_map.py

```diff
@@ -0,0 +1,138 @@
+    transformed = counts.transform_coords(
+        ['Q_parallel', 'Q_perpendicular', 'Q'],
+        graph={
+            'Q_parallel': _make_projection_kernel(q_projection.parallel),
+            'Q_perpendicular': _make_projection_kernel(q_projection.perpendicular),
+            'Q': lambda sample_table_momentum_transfer: sc.norm(
+                sample_table_momentum_transfer
+            ),
+        },
+        keep_inputs=False,
+    )
+    if transformed.bins is not None:
+        transformed.bins.coords['a3'] = sc.bins_like(
+            transformed, transformed.coords['a3']
+        )
+        transformed = transformed.bins.concat()
+
+    return CountsWithQMapCoords[RunType](transformed)
+
+
+def _make_projection_kernel(vector: sc.Variable) -> Callable[..., sc.Variable]:
+    normal = vector / sc.norm(vector)
+
+    def projection(sample_table_momentum_transfer: sc.Variable) -> sc.Variable:
+        return sc.dot(sample_table_momentum_transfer, normal)
+
+    return projection
+
+
+def default_q_projection() -> QProjection:
+    return QProjection(
+        parallel=sc.vector(value=[0, 0, 1]), perpendicular=sc.vector(value=[1, 0, 0])
+    )
+
+
+def make_q_map(
+    events: sc.DataArray,
+    q_parallel_bins: int | sc.Variable,
+    q_perpendicular_bins: int | sc.Variable,
+    sample_rotation_bins: int | sc.Variable,
+) -> Box:
+    """Build a figure with a 2D Q-map and a 1D slice for a range of Q values."""
+    import ipywidgets as ipw
+
+    def make_q_hist(da: sc.DataArray) -> sc.DataArray:
+        return da.hist(
+            Q_perpendicular=q_perpendicular_bins, Q_parallel=q_parallel_bins
+        ).rename(Q_perpendicular=r'$Q_{\perp}$', Q_parallel='$Q_{||}$')
+
+    def make_q_slice(da: sc.DataArray, q_range: tuple[float, float]) -> sc.DataArray:
+        unit = da.bins.coords['Q'].unit
+        lo = q_range[0] * unit
+        hi = q_range[1] * unit
+        sliced = da.bins['Q', lo:hi]
+        try:
+            return sliced.hist(a3=sample_rotation_bins).rename(a3='sample_rotation')
+        except ValueError as err:
+            if "empty data range" in err.args[0].lower():
+                return _empty_angle_array_like(sliced)
+            raise
+
+    q_lo = events.bins.coords['Q'].min().value
+    q_hi = events.bins.coords['Q'].max().value
+    slider = ipw.FloatRangeSlider(
+        value=(q_lo, q_hi), min=q_lo, max=q_hi, step=(q_hi - q_lo) / 100
+    )
+
+    slider_node = pp.widget_node(slider)
+    input_node = pp.Node(events)
+    q_map_node = pp.Node(make_q_hist, input_node)
+    q_slice_node = pp.Node(make_q_slice, input_node, slider_node)
+    q_map_fig = pp.imagefigure(q_map_node, aspect='equal', norm='log')
+    q_slice_fig = pp.linefigure(q_slice_node)
+
+    lo_circle = _draw_circle(q_map_fig.ax, q_lo)
+    hi_circle = _draw_circle(q_map_fig.ax, q_hi)
+
+    def update_circle(q_range: tuple[float, float]) -> None:
+        lo_circle.radius, hi_circle.radius = q_range
+        q_map_fig.fig.canvas.draw_idle()
```

Comment:
```suggestion
        q_map_fig.canvas.draw()
```
This sfould also work